File: ui/src/model/song.ts
import { IBase, IBaseSearchParams } from "@allape/gocrud";
import { ICollection } from "./collection.ts";
import { ILyrics } from "./lyrics.ts";

export interface ISong extends IBase {
  name: string;
  filename: string;
  cover: string;
  digest: string;
  mime: string;
  ffprobeInfo: string;
  description: string;
  index: number;
}

export interface ISongSearchParams extends IBaseSearchParams {
  like_name?: string;
  in_id?: ISong["id"][];
  collectionId?: ICollection["id"];
  orderBy_index?: "desc" | "asc";
  orderBy_createdAt?: "desc" | "asc";
  orderBy_updatedAt?: "desc" | "asc";
}

export interface ISongLyrics {
  songId: ISong["id"];
  lyricsId: ILyrics["id"];
  createdAt: string;
}

File: model/song.go
package model

import (
	"github.com/allape/gocrud"
	"time"
)

type Song struct {
	gocrud.Base
	Name        string `json:"name"`
	Filename    string `json:"filename"`
	Cover       string `json:"cover"`
	Digest      string `json:"digest"`
	MIME        string `json:"mime"`
	FFProbeInfo string `json:"ffprobeInfo"`
	Description string `json:"description"`
	Index       int32  `json:"index" gorm:"default:0"`
}

type SongLyrics struct {
	SongID    gocrud.ID `json:"songId"`
	LyricsID  gocrud.ID `json:"lyricsId"`
	CreatedAt time.Time `json:"createdAt" gorm:"autoCreateTime;<-:create"`
}
